cups/file.c

```diff
@@ -335,6 +335,165 @@ _cupsFileCheckFilter(
 #endif /* !_WIN32 */
 
 
+/*
+ * '_cupsFileGetConfAndComments()' - Get line and comments from a configuration file.
+ */
+
+char *					/* O  - Line read or @code NULL@ on end of file or error */
+_cupsFileGetConfAndComments(cups_file_t *fp,	/* I  - CUPS file */
+	char        *buf,	/* O  - String buffer */
+	size_t      buflen,	/* I  - Size of string buffer */
+	char        **value,	/* O  - Pointer to value */
+	int         *linenum)	/* IO - Current line number */
+{
+  char	*ptr;				/* Pointer into line */
+
+
+ /*
+  * Range check input...
+  */
+
+  DEBUG_printf(("2cupsFileGetConfAndComments(fp=%p, buf=%p, buflen=" CUPS_LLFMT
+                ", value=%p, linenum=%p)", (void *)fp, (void *)buf, CUPS_LLCAST buflen, (void *)value, (void *)linenum));
+
+  if (!fp || (fp->mode != 'r' && fp->mode != 's') ||
+      !buf || buflen < 2 || !value)
+  {
+    if (value)
+      *value = NULL;
+
+    return (NULL);
+  }
+
+ /*
+  * Read the next line...
+  */
+
+  *value = NULL;
+
+  while (cupsFileGets(fp, buf, buflen))
+  {
+    (*linenum) ++;
+
+    /*
+     * Remove the inline comment...
+     */
+    if ((ptr = strchr(buf, '#')) != NULL)
+    {
+      /*
+       * Check if the '#' character is escaped by a backslash...
+       */
+      if (ptr != buf && ptr[-1] == '\\')
+      {
+        /*
+         * Remove the backslash and continue searching for unescaped '#'...
+         */
+        _cups_strcpy(ptr-1, ptr);
+        ptr = strchr(ptr+1, '#');
+      }
+
+      /*
+       * Find the last non-whitespace character before the unescaped '#' character...
+       */
+      while (ptr > buf)
+      {
+        if (!_cups_isspace(ptr[-1]))
+        {
+          /*
+           * Null-terminate the string after the last non-whitespace character...
+           */
+          *ptr = '\0';
+          break;
+        }
+
+        /*
+         * Check if the '#' character is escaped by a backslash...
+         */
+        if (ptr != buf && *(ptr-1) == '\\')
+        {
+          /*
+           * Remove the backslash and continue searching for unescaped '#'
+           */
+          _cups_strcpy(ptr-1, ptr);
+          ptr = strchr(ptr+1, '#');
+        }
+        else
+        {
+          ptr--;
+        }
+      }
+    }
+
+	/*
+	 * Strip leading whitespace...
+	 */
+
+    for (ptr = buf; _cups_isspace(*ptr); ptr ++);
+
+    if (ptr > buf)
+      _cups_strcpy(buf, ptr);
+
+    /*
```

Comment:
Please change the comment here - now we don't return only comment, but in case of empty line as well.